src/ucp/proto/rndv.c

```diff
@@ -674,32 +680,40 @@ static ucs_status_t ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req,
 UCS_PROFILE_FUNC_VOID(ucp_rndv_recv_frag_put_completion, (self, status),
                       uct_completion_t *self, ucs_status_t status)
 {
-    ucp_request_t *freq     = ucs_container_of(self, ucp_request_t,
-                                               send.state.uct_comp);
-    ucp_request_t *req      = freq->send.rndv_put.sreq;
-    ucp_request_t *rndv_req = (ucp_request_t*)freq->send.rndv_put.remote_request;
+    ucp_request_t *freq              = ucs_container_of(self, ucp_request_t,
+                                                        send.state.uct_comp);
+    ucs_ptr_map_key_t rreq_remote_id = freq->send.rndv_put.rreq_remote_id;
+    int is_put_proto                 = (rreq_remote_id == UCP_REQUEST_ID_INVALID);
+    ucp_request_t *req               = freq->send.rndv_put.sreq;
+    ucp_request_t *rndv_req;
 
     ucs_trace_req("freq:%p: recv_frag_put done. rreq:%p ", freq, req);
 
     /* release memory descriptor */
     ucs_mpool_put_inline((void *)freq->send.mdesc);
 
     /* rndv_req is NULL in case of put protocol */
-    if (rndv_req != NULL) {
+    if (!is_put_proto) {
+        rndv_req = ucp_worker_get_request_by_id(freq->send.ep->worker,
+                                                rreq_remote_id);
         /* pipeline recv get protocol */
         rndv_req->send.state.dt.offset += freq->send.length;
 
         /* send ATS for fragment get rndv completion */
         if (rndv_req->send.length == rndv_req->send.state.dt.offset) {
             ucp_rkey_destroy(rndv_req->send.rndv_get.rkey);
             ucp_rndv_req_send_ats(rndv_req, req,
-                                  rndv_req->send.rndv_get.remote_request, UCS_OK);
+                                  rndv_req->send.rndv_get.remote_req_id,
+                                  UCS_OK);
         }
     }
 
     req->recv.tag.remaining -= freq->send.length;
     if (req->recv.tag.remaining == 0) {
         ucp_request_complete_tag_recv(req, UCS_OK);
+        if (!is_put_proto) {
+            ucp_worker_del_request_id(freq->send.ep->worker, rreq_remote_id);
```

Comment:
can we add leak check in ucs_ptr_map_destroy() or when ucp_worker_destroy() calls it?
e.g warn if any entries in that hash